airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/components/debezium/DebeziumProducer.kt

```diff
@@ -0,0 +1,282 @@
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog
+import io.airbyte.protocol.models.v0.SyncMode
+import io.debezium.engine.ChangeEvent
+import io.debezium.engine.DebeziumEngine
+import io.debezium.engine.format.Json
+import io.debezium.spi.common.ReplacementFunction
+import io.debezium.storage.file.history.FileSchemaHistory
+import java.lang.reflect.Method
+import java.time.Duration
+import java.util.*
+import java.util.regex.Pattern
+import org.apache.kafka.connect.connector.Connector
+import org.apache.kafka.connect.runtime.standalone.StandaloneConfig
+import org.apache.kafka.connect.source.SourceRecord
+import org.apache.kafka.connect.storage.FileOffsetBackingStore
+import org.codehaus.plexus.util.StringUtils
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+/** [ProducerComponent] for Debezium. Essentially a wrapper around [DebeziumEngine]. */
+class DebeziumProducer
+private constructor(
+    debeziumProperties: Properties,
+    boundChecker: BoundChecker,
+    consumer: ConsumerComponent<DebeziumRecord>,
+    notifyStop: () -> Unit,
+    private val initialState: DebeziumState,
+) : ProducerComponent<DebeziumState> {
+
+    private val stateFilesAccessor = StateFilesAccessor()
+
+    init {
+        // Write offset and schema files.
+        stateFilesAccessor.writeOffset(initialState.offset)
+        initialState.schema.ifPresent(stateFilesAccessor::writeSchema)
+        // Add missing offset and schema history file path properties.
+        debeziumProperties.setProperty(
+            StandaloneConfig.OFFSET_STORAGE_FILE_FILENAME_CONFIG,
+            stateFilesAccessor.offsetFilePath.toString(),
+        )
+        if (initialState.schema.isPresent) {
+            debeziumProperties.setProperty(
+                FileSchemaHistory.FILE_PATH.name(),
+                stateFilesAccessor.schemaFilePath.toString(),
+            )
+        }
+    }
+
+    private val engine: DebeziumEngine<ChangeEvent<String?, String?>> =
+        DebeziumEngine.create(Json::class.java)
+            .using(debeziumProperties)
+            // Event handler.
+            .notifying { event ->
+                if (event.value() == null) {
+                    // Debezium outputs a tombstone event that has a value of null. This is an
+                    // artifact of how it interacts with kafka. We want to ignore it. More on the
+                    // tombstone:
+                    // https://debezium.io/documentation/reference/2.2/transformations/event-flattening.html
+                    return@notifying
+                }
+                val record = DebeziumRecord(Jsons.deserialize(event.value()))
+                // Heartbeat events aren't useful to the consumer, they're used for bound checks.
+                if (!record.isHeartbeat) {
+                    consumer.accept(record)
+                    if (consumer.shouldCheckpoint()) {
+                        // Stop if we need to checkpoint.
+                        notifyStop()
```

Comment:
Have we ported over the debezium shutdown procedure when we shut down debezium? 

We need to do this to synchronize the state in the offset file with whatever debezium has already fetched. Airbyte may have processed 10,000 records, but debezium could buffered more